src/contracts/core/Slasher.sol

```diff
@@ -1,102 +1,368 @@
 
-    function middlewareTimesLength(address) external view returns (uint256) {}
+        IOperatorSetManager.OperatorSet memory operatorSet =
+            IOperatorSetManager.OperatorSet({avs: msg.sender, id: operatorSetID});
+        bytes32 operatorSetHash = _hashOperatorSet(operatorSet);
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            IStrategy strategy = strategies[i];
+            uint32 requestedSlashedBipsBefore = requestedSlashedBips[operator][strategy][epoch][operatorSetHash];
+            int32 requestedSlashedBipsAfter = int32(requestedSlashedBipsBefore) + bipsToModify;
+            // don't allow underflow
+            if (requestedSlashedBipsAfter < 0) {
+                bipsToModify = -int32(requestedSlashedBipsBefore);
+                requestedSlashedBipsAfter = 0;
+            }
+            requestedSlashedBips[operator][strategy][epoch][operatorSetHash] = uint32(requestedSlashedBipsAfter);
 
-    function getMiddlewareTimesIndexStalestUpdateBlock(address, uint32) external view returns (uint32) {}
+            SlashingRequest memory slashingRequest = slashingRequests[operator][strategy][epoch];
+            // if this is the first request, increment the lastCreatedSlashingRequestId and set the id
+            if (slashingRequest.id == 0) {
+                uint32 requestId = slashingRequestIds[operator][strategy].lastCreatedSlashingRequestId + 1;
+                slashingRequestIds[operator][strategy].lastCreatedSlashingRequestId = requestId;
+                slashingRequest.id = requestId;
+            }
 
-    function getMiddlewareTimesIndexServeUntilBlock(address, uint32) external view returns (uint32) {}
+            // pending slashing rate is modified by the bips to modify * the slashable bips
+            // when divided by BIPS_FACTOR**2, this will give the actual proportion of the
+            // stake is being modified. this is done for accuracy when slashable bips are small
+            slashingRequest.slashingRate = uint64(
+                int64(slashingRequest.slashingRate)
+                    + int64(bipsToModify)
+                        * int64(uint64(operatorSetManager.getSlashableBips(operator, operatorSet, strategy, epoch)))
+            );
+            slashingRequests[operator][strategy][epoch] = slashingRequest;
+        }
 
-    function operatorWhitelistedContractsLinkedListSize(address) external view returns (uint256) {}
+        emit RequestedBipsToSlashModified(epoch, operator, operatorSet, strategies, bipsToModify);
+    }
 
-    function operatorWhitelistedContractsLinkedListEntry(
-        address,
-        address
-    ) external view returns (bool, uint256, uint256) {}
+    /// @notice Returns the epoch for lookup in the slashing request history and whether
+    /// there were any slashing requests that happened before or at the given epoch
+    function _getLookupEpoch(address operator, IStrategy strategy, uint32 epoch) internal view returns (uint32, bool) {
```

Comment:
IMO this is abysmal naming. in particular this is looking up the most recent epoch where slashing occurred for the (operator, strategy) tuple, which is at or prior to the `epoch` -- *and* returning false if no such value is found